### DISCUSSION TITLE
my _document.js stopped working after upgrading to latest version of next.js

### DISCUSSION THREAD

**Discussion by Vishakh-vcc** [vercel/next.js#51366] [Help]
### Summary

I upgraded to latest version of next js and finally my code stopped working in the _document.js file .... attaching my _document.js file and also error which I am facing
Kindly provide me a latest working file 

_document.js file :
===============
import React from 'react';
import Document, { Html, Head, Main, NextScript } from 'next/document';
import { StyleProvider, ThemeProvider, StyleTags } from 'vcc-ui';
import { getRenderer } from '../vcc-ui.config.js';
import { theme } from '../theme.js'
import { ServerStyleSheet } from 'styled-components';

export default class MyDocument extends Document {
  static getInitialProps(props) {
    const serverRenderer = getRenderer();
    const page = props.renderPage(
      App =>
        function document(props) {
          return (
            <StyleProvider renderer={serverRenderer}>
              <ThemeProvider theme={theme}>
                <App {...props} />
              </ThemeProvider>
            </StyleProvider>
          );
        }
    );

    return { ...page, renderer: serverRenderer };
  }

  render() {
    return (
      <Html>
        <Head>
          <StyleTags renderer={this.props.renderer} />
        </Head>
        <body>
          <Main />
          <NextScript />
        </body>
      </Html>
    );
  }
}

Error message : 
===========
Error:  MyDocument.getInitialProps()" should resolve to an object with a "html" prop set with a valid html string

### Additional information

```js
_document.js file :
===============
import React from 'react';
import Document, { Html, Head, Main, NextScript } from 'next/document';
import { StyleProvider, ThemeProvider, StyleTags } from 'vcc-ui';
import { getRenderer } from '../vcc-ui.config.js';
import { theme } from '../theme.js'
import { ServerStyleSheet } from 'styled-components';

export default class MyDocument extends Document {
  static getInitialProps(props) {
    const serverRenderer = getRenderer();
    const page = props.renderPage(
      App =>
        function document(props) {
          return (
            <StyleProvider renderer={serverRenderer}>
              <ThemeProvider theme={theme}>
                <App {...props} />
              </ThemeProvider>
            </StyleProvider>
          );
        }
    );

    return { ...page, renderer: serverRenderer };
  }

  render() {
    return (
      <Html>
        <Head>
          <StyleTags renderer={this.props.renderer} />
        </Head>
        <body>
          <Main />
          <NextScript />
        </body>
      </Html>
    );
  }
}



Error message : 
===========
Error:  MyDocument.getInitialProps()" should resolve to an object with a "html" prop set with a valid html string
```

**Comment by icyJoseph** [comment]
@Vishakh-vcc I think you can do this:

```tsx
import Document, {
  Html,
  Head,
  Main,
  NextScript,
  DocumentContext,
} from "next/document";
import { StyleTags, renderStyles } from "vcc-ui";
import { getRenderer } from "../../vcc-ui.config.js";

export default class MyDocument extends Document {
  static async getInitialProps(ctx) {
    const serverRenderer = getRenderer();

    const originalRenderPage = ctx.renderPage;

    ctx.renderPage = () =>
      originalRenderPage({
        enhanceApp(App) {
          return (props) => {
            return <App {...props} renderer={serverRenderer} />;
          };
        },
      });

    const initialProps = await Document.getInitialProps(ctx);
    const styles = renderStyles(serverRenderer);

    return {
      ...initialProps,
      styles: [initialProps.styles, styles],
      renderer: serverRenderer,
    };
  }

  render() {
    return (
      <Html>
        <Head>
          <StyleTags renderer={this.props.renderer} />
        </Head>
        <body>
          <Main />
          <NextScript />
        </body>
      </Html>
    );
  }
}

```

And then your `_app`:

```jsx
import { StyleProvider, ThemeProvider } from "vcc-ui";
import volvo from "vcc-ui/dist/themes/volvo";

function MyApp({
  Component,
  pageProps,
  renderer,
}) {
  return (
    <StyleProvider renderer={renderer}>
      <ThemeProvider theme={volvo}>
        <Component {...pageProps} />
      </ThemeProvider>
    </StyleProvider>
  );
}

export default MyApp;

```

**Reply by icyJoseph** [reply]
@Vishakh-vcc 

What `vcc-ui` version have you got installed?

**Reply by icyJoseph** [reply]
Also the renderer I used I got like this:

```js
// vcc-ui.config.js
import { styleRenderer } from "vcc-ui";

export function getRenderer() {
  const renderer = styleRenderer();
  renderer.renderStatic(
    {
      margin: 0,
      padding: 0
    },
    "body"
  );
  return renderer;
}
```

**Reply by Vishakh-vcc** [reply]
@icyJoseph  my vcc-ui version is : 2.10.1, if I try to install higher version again I get the below kind of 'feta' error 

![image](https://github.com/vercel/next.js/assets/71757941/59edf1f7-d495-4469-8966-2079c6288630)


This is my renderer:
// vcc-ui.config.js
import { styleRenderer } from 'vcc-ui';
//import volvo from 'vcc-ui/lib/themes/volvo';

// export const volvoTheme = {
//   ...volvo,
//   fontsPath: '/static/vcc-ui/fonts/',
//   logoImagesPath: '/static/vcc-ui/images/',
// };

export function getRenderer() {
  const renderer = styleRenderer();
  renderer.renderStatic(
    {
      margin: 0,
      height: "100%",
      fontFamily: "Volvo Novum Light, sans-serif"
    },
    'body'
  );

  return renderer;
}

**Reply by icyJoseph** [reply]
You mean the `lstat` error from fela?

**Reply by icyJoseph** [reply]
@Vishakh-vcc what Node.js version have you got? 

I am using Node 18, Next@13.4.3, and vcc-ui@4.4.0, and it works well. 

<img width="1206" alt="Screenshot 2023-06-20 at 19 05 51" src="https://github.com/vercel/next.js/assets/21013447/478b7bca-8f41-4f8e-b66d-0a3b67ec4e96">

Using the code I gave you.

**Reply by Vishakh-vcc** [reply]
@icyJoseph 

versions:
Node : v20.3.0 
"next": "13.4.5"
"vcc-ui": "^2.10.1"

when I change my vcc-ui to latest version it suggests me to change few source codes in multiple files which I wanted to avoid as of now , hence seeking your help

**Reply by icyJoseph** [reply]
I switched to vcc-ui@2.26.0 and Node 20.3.0, and my code works fine. It works on dev and production mode.

I am starting to think this is a Windows issue. https://github.com/nodejs/node/issues/43439

Could you remove `node_modules` and `.next`, and reinstall everything again with `npm install` or `yarn`, which ever you use.

**Comment by Vishakh-vcc** [comment]
Thanks for suggestion, I made changes as suggested by you , I am getting below error after that , how can I proceed from here


***@***.***

**Reply by Vishakh-vcc** [reply]
![errorGit](https://github.com/vercel/next.js/assets/71757941/35d4a544-6ec6-4534-b359-12bf886c4d35)


Thanks for suggestion, I made changes as suggested by you , I am getting below error after that , how can I proceed from here

**Comment by Vishakh-vcc** [comment]
@icyJoseph  Thanks for suggestion, I made changes as suggested by you , I am getting below error after that , how can I proceed from here

![errorGit](https://github.com/vercel/next.js/assets/71757941/5a5ceede-2b48-4b32-8a03-adcfdd9a6f80)

**Comment by MiftahClassifieds** [comment]
JUST rename your _document.js file to _document.cjs and all issues will resolved!

**Reply by Vishakh-vcc** [reply]
@MiftahClassifieds 
When I did this I am getting the below error again :

Error: UNKNOWN: unknown error, lstat 'C:\Users\vramana1\OneDrive - Volvo Cars\Documents\GitHub\icoe-apim-devloper-portal\node_modules\fela-dom\lib'

Error: The "useFela" hook can only be used inside a "RendererProvider"

Have you encountered this error

**Comment by svignara** [comment]
> JUST rename your _document.js file to _document.cjs and all issues will resolved!

@MiftahClassifieds Thanks! This fixed the issue for me, but why does this work, and why doesn't it work when it's `document.js` anymore?

**Reply by icyJoseph** [reply]
Most likely had module type in package.json. 

- https://nodejs.org/api/packages.html#packagejson-and-file-extensions
